evm-contracts/src/v0.8/dev/ValidatorProxy.sol

```diff
@@ -0,0 +1,251 @@
+   * the call was received.
+   * @param previousRoundId uint256
+   * @param previousAnswer int256
+   * @param currentRoundId uint256
+   * @param currentAnswer int256
+   * @return bool
+   */
+  function validate(
+    uint256 previousRoundId,
+    int256 previousAnswer,
+    uint256 currentRoundId,
+    int256 currentAnswer
+  )
+    external
+    override
+    returns (
+      bool
+    )
+  {
+    address currentAggregator = s_currentAggregator.target;
+    address proposedAggregator = s_proposedAggregator;
+    require(msg.sender == currentAggregator || msg.sender == proposedAggregator, "Not a configured aggregator");
+    // If the aggregator is still in proposed state, emit an event and don't push to any validator.
+    // This is to confirm that `validate` is being called prior to upgrade.
+    if (msg.sender == proposedAggregator) {
+      emit ProposedAggregatorValidateCall(
+        proposedAggregator,
+        previousRoundId,
+        previousAnswer,
+        currentRoundId,
+        currentAnswer
+      );
+      return true;
+    }
+
+    // Send the validate call to the current validator
+    ProxyConfiguration memory currentValidator = s_currentValidator;
+    require(s_currentValidator.target != address(0), "No validator set");
+    AggregatorValidatorInterface(currentValidator.target).validate(
+      previousRoundId,
+      previousAnswer,
+      currentRoundId,
+      currentAnswer
+    );
+    // If there is a new proposed validator, send the validate call to that validator also
+    if (currentValidator.hasNewProposal) {
+      AggregatorValidatorInterface(s_proposedValidator).validate(
+        previousRoundId,
+        previousAnswer,
+        currentRoundId,
+        currentAnswer
+      );
+    }
+    return true;
+  }
+
+  /** AGGREGATOR CONFIGURATION FUNCTIONS **/
+
+  /**
+   * @notice Propose an aggregator
+   * @dev A zero address can be used to unset the proposed aggregator. Only owner can call.
+   * @param proposed address
+   */
+  function proposeNewAggregator(
+    address proposed
+  )
+    external
+    onlyOwner()
+  {
+    s_proposedAggregator = proposed;
+    // If proposed is zero address, hasNewProposal = false
+    s_currentAggregator.hasNewProposal = (proposed != address(0));
+    emit AggregatorProposed(proposed);
+  }
+
+  /**
+   * @notice Upgrade the aggregator by setting the current aggregator as the proposed aggregator.
+   * @dev Must have a proposed aggregator. Only owner can call.
+   */
+  function upgradeAggregator()
+    external
+    onlyOwner()
+  {
+    // Get configuration in memory
+    ProxyConfiguration memory current = s_currentAggregator;
+    address previous = current.target;
+    address proposed = s_proposedAggregator;
+
+    // Perform the upgrade
+    require(current.hasNewProposal == true, "No proposal");
+    current.target = proposed;
+    current.hasNewProposal = false;
+
+    s_currentAggregator = current;
```

Comment:
I would prefer initializing a new struct here since it *forces* you to not forget any fields

```suggestion
    s_currentAggregator = ProxyConfiguration({
      target: proposed,
      hasNewProposal: false
    });
```